auction/contracts/AuctionHouse.sol

```diff
@@ -0,0 +1,338 @@
+        uint currentTime = block.timestamp;
+        //start action if minimal price is met
+        if (currentAuction.buyer == address(0x0)) {//no bid at all
+            // set endTime if it's not set
+            if (currentAuction.endTime == 0){
+                auctions[_auctionId].endTime = currentTime + aucData.duration;
+            }
+            require(newAmount >= currentAuction.minimalPrice, "bid can't be less than minimal price");
+        } else {//there is bid in auction
+            require(currentAuction.buyer != newBuyer, "already have an outstanding bid");
+            uint256 minAmount = getMinimalNextBid(_auctionId);
+            require(newAmount >= minAmount, "bid amount too low");
+        }
+        reserveValue(
+            currentAuction.buyAsset, 
+            currentAuction.buyer, 
+            newBuyer, 
+            getBidTotalAmount(currentAuction.lastBid, currentAuction.protocolFee), 
+            getBidTotalAmount(bid, currentAuction.protocolFee)
+        );
+        auctions[_auctionId].lastBid = bid;
+        auctions[_auctionId].buyer = newBuyer;
+
+        // extends auction time if it about to end
+        if (currentAuction.endTime - currentTime < EXTENSION_DURATION) {
+            auctions[_auctionId].endTime = currentTime + EXTENSION_DURATION;
+        }
+        emit BidPlaced(_auctionId, newBuyer, bid, auctions[_auctionId].endTime);
+    }
+
+    /// @dev reserves new bid and returns the last one if it exists
+    function reserveValue(LibAsset.AssetType memory _buyAssetType, address oldBuyer, address newBuyer, uint oldAmount, uint newAmount) internal {
+        LibAsset.Asset memory transferAsset;
+        if (oldBuyer != address(0x0)) {//return oldAmount to oldBuyer
+            transferAsset = LibAsset.Asset(_buyAssetType, oldAmount);
+            transfer(transferAsset, address(this), oldBuyer, TO_LOCK, UNLOCK);
+        }
+        transferAsset = LibAsset.Asset(_buyAssetType, newAmount);
+        if (transferAsset.assetType.assetClass == LibAsset.ETH_ASSET_CLASS) {
+            if (msg.value > newAmount) {//more ETH than need
+                address(newBuyer).transferEth(msg.value - newAmount);
+            }
+        } else {
+            transfer(transferAsset, newBuyer, address(this), TO_LOCK, LOCK);
+            setApproveForTransferProxy(transferAsset);
+        } 
+    }
+
+    /// @dev returns the minimal amount of the next bid (without fees)
+    function getMinimalNextBid(uint _auctionId) public view returns (uint minBid){
+        Auction storage currentAuction = auctions[_auctionId];
+
+        if (currentAuction.buyer == address(0x0)) {
+            minBid = currentAuction.minimalPrice;
+        } else {
+            minBid = currentAuction.lastBid.amount + currentAuction.minimalStep;
+        }
+    }
+
+    /// @dev returns true if auction exists, false otherwise
+    function checkAuctionExistence(uint _auctionId) public view returns (bool){
```

Comment:
you have a similar method checkAuctionInProgress and it includes "require" inside.
this method is always called with "require", thus we can include "require" here too

Reply:
need to think about it

Reply:
this is public method for external users so require here is not needed i think
tho i deleted isFinilized() method, coz it was almost identical to checkAuctionExistence